packages/core/src/components/forms/numericInput.tsx

```diff
@@ -0,0 +1,407 @@
+                iconName={iconName}
+                intent={this.props.intent}
+                key={key}
+                onBlur={this.handleButtonBlur}
+                onClick={onClick}
+                onFocus={this.handleButtonFocus}
+                onKeyDown={onKeyDown}
+            />
+        );
+    }
+
+    private inputRef = (input: HTMLInputElement) => {
+        this.inputElement = input;
+    }
+
+    private handleDecrementButtonClick = (e: React.MouseEvent<HTMLInputElement>) => {
+        this.updateValue(-1, e);
+    }
+
+    private handleIncrementButtonClick = (e: React.MouseEvent<HTMLInputElement>) => {
+        this.updateValue(+1, e);
+    }
+
+    private handleButtonFocus = () => {
+        this.setState({ isButtonGroupFocused: true });
+    }
+
+    private handleButtonBlur = () => {
+        this.setState({ isButtonGroupFocused: false });
+    }
+
+    private handleButtonKeyDown =
+            (e: React.KeyboardEvent<HTMLInputElement>, onClick: React.MouseEventHandler<HTMLInputElement>) => {
+        if (e.keyCode === Keys.SPACE) {
+            // prevent the page from scrolling (this is the default browser
+            // behavior for shift + space or alt + space).
+            e.preventDefault();
+
+            // trigger a click event to update the input value appropriately,
+            // based on the active modifier keys.
+            const fakeClickEvent = {
+                altKey: e.altKey,
+                currentTarget: e.currentTarget,
+                shiftKey: e.shiftKey,
+                target: e.target,
+            };
+            onClick(fakeClickEvent as React.MouseEvent<HTMLInputElement>);
+        }
+    }
+
+    private handleInputFocus = () => {
+        this.setState({ isInputGroupFocused: true });
+    }
+
+    private handleInputBlur = () => {
+        this.setState({ isInputGroupFocused: false });
+        this.handleDone();
+    }
+
+    private handleInputKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+        if (this.props.disabled || this.props.readOnly) {
+            return;
+        }
+
+        const { keyCode } = e;
+
+        let direction: number;
+
+        if (keyCode === Keys.ENTER) {
+            this.handleDone();
+            return;
+        } else if (keyCode === Keys.ARROW_UP) {
+            direction = 1;
+        } else if (keyCode === Keys.ARROW_DOWN) {
+            direction = -1;
+        } else {
+            return;
+        }
+
+        // we'd like to select the field contents after running the code in this
+        // onKeyDown handler, as a UX nicety. without e.preventDefault, some
+        // hotkeys (e.g. shift + up/down, alt + up/down) will clear the selection,
+        // resulting in an inconsistent or unintuitive experience.
+        e.preventDefault();
+
+        this.updateValue(direction, e);
+    }
+
+    private handleDone = () => {
+        if (this.props.onConfirm != null) {
+            this.props.onConfirm(this.state.value);
```

Comment:
Done. Though in this particular case, I think the logic is cleaner as is, since I need to null-check `this.props.onConfirm` anyway:

```
private handleDone = () => {
    Utils.safeInvoke(this.props.onConfirm, this.state.value);

    if (this.props.onConfirm != null) {
        const { min, max } = this.props;
        const currValue = this.state.value;
        const nextValue = (currValue.length > 0) ? this.getAdjustedValue(currValue, /* delta */ 0, min, max) : "";
        this.setState({ value: nextValue });
    }
}
```